src/data/waffleFlags.test.js

```diff
@@ -0,0 +1,167 @@
+    const fn = jest.fn();
+    const then = jest.fn();
+
+    executePossiblyAsyncFunction(fn, then)
+      .then(() => {
+        expect(fn).toHaveBeenCalledTimes(1);
+        expect(then).toHaveBeenCalledTimes(1);
+      });
+  });
+
+  it('should execute as a AsyncFunction', async (done) => {
+    const fnInnerMock = jest.fn();
+    const then = jest.fn();
+    const fn = async () => { fnInnerMock(); };
+
+    executePossiblyAsyncFunction(fn, then)
+      .then(() => {
+        expect(fnInnerMock).toHaveBeenCalledTimes(1);
+        expect(then).toHaveBeenCalledTimes(1);
+      })
+      .then(done);
+  });
+});
+
+/**
+ * Set our JSDOM Window and Document Location
+ *
+ * **NOTE**: You *MUST* include your tests async DoneCallback as the final action of your `perform`'s closure.
+ * failure to do so will lead to test timeouts.
+ *
+ * @param {string} url The URL we intend to go to, this must be within the domain of `window.origin`
+ * @param {Function} perform
+ * @see window.origin
+ * @see history.pushState
+ * @throws {SecurityError} if not within the domain of `window.origin`
+ * @return Promise<void>
+ */
+const performWithModifiedJSDOMLocation = (url, perform) => {
+  const lastValue = window.location.href;
+
+  /* eslint-disable-next-line no-restricted-globals */ // We need this for some test manipulation (history object)
+  history.pushState(history.state, null, new URL(url));
+
+  expect(document.location.href).toBe(url);
+  expect(window.location.href).toBe(url);
+
+  const cleanup = () => {
+    /* eslint-disable-next-line no-restricted-globals */ // We need this for some test manipulation (history object)
+    history.pushState(history.state, null, new URL(lastValue));
+  };
+
+  return executePossiblyAsyncFunction(perform, cleanup);
+};
+
+/**
+ * Perform a test with modified waffle flags.
+ *
+ * This function ensures the state is returned to its default.
+ *
+ * **NOTE**: You *MUST* include your tests async DoneCallback as the final action of your `perform`'s closure.
+ * failure to do so will lead to test timeouts.
+ *
+ * @param {{}} flags
+ * @param {() => void} perform
+ * @return Promise<void>
+ */
+export function performWithModifiedWaffleFlags(flags, perform) {
+  const initialConfig = getConfig().WAFFLE_FLAGS;
+  mergeConfig({ WAFFLE_FLAGS: flags });
+
+  const cleanup = () => {
+    mergeConfig({ WAFFLE_FLAGS: initialConfig });
+  };
+
+  return executePossiblyAsyncFunction(perform, cleanup);
+}
+
+describe('getWaffleFlags', () => {
+  it('should default to document.location when empty', () => {
+    // We have to use the Window's origin, otherwise: SecurityError: replaceState cannot update...
+    //   This was WAY TOO HARD FOUGHT.
+    const testLocation = `${window.origin}/dox.asp?${WAFFLE_PREFIX}xyzzy=on`;
+
+    return performWithModifiedJSDOMLocation(testLocation, () => {
+      const result = processUrlWaffleFlags();
+      expect(result).toStrictEqual({ xyzzy: true });
+    });
```

Comment:
No, the machoism for handling history in Jest doesnt work right as it mimics a full DOM via JSDOM. Wes and i tried some 7 different forms, none took. This is why we had to work with the history scoping mechanism as well because it throws various `SecurityError: `'s, such as the one depper in the stack called `replaceState` even though we never touch it directly, its minicing the Mozilla Docs (i assume) and such.

Reply:
Also theres needed cleanup, to reset the DOM for the next call, this requires the caller to know nothing of that, and it 'just works'.